### PR TITLE
WIP: Synchronize range data for multiple sensors

### PR DESCRIPTION
**Pull request by gaschler**
Not ready for review.

FIXES=#910

### CODE REVIEW

File: cartographer/mapping/internal/3d/local_trajectory_builder_3d_test.cc [hunk 1]
```diff
@@ -35,6 +35,8 @@ namespace cartographer {
 namespace mapping {
 namespace {
 
+const char kSensorId[] = "sensor_id";
```
Comment (pifon2a) [inline cartographer/mapping/internal/3d/local_trajectory_builder_3d_test.cc]: nit: constexpr

File: cartographer/mapping/internal/range_data_collator.cc [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+#include "cartographer/mapping/local_slam_result_data.h"
+#include "glog/logging.h"
+
+namespace cartographer {
+namespace mapping {
+
+sensor::TimedPointCloudOriginData RangeDataCollator::AddRangeData(
+    const std::string& sensor_id,
+    const sensor::TimedPointCloudData& timed_point_cloud_data) {
+  CHECK(expected_sensor_ids_.count(sensor_id) != 0);
```
Comment (pifon2a) [inline cartographer/mapping/internal/range_data_collator.cc]: Is this CHECK really needed? Can this situation occur?

File: cartographer/mapping/internal/range_data_collator.cc [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    current_start_ = current_end_;
+    // When we have two messages of the same sensor, move forward the older of
+    // the two (do not send out current).
+    current_end_ = id_to_pending_data_.at(sensor_id).time;
+    auto result = CropAndMerge();
+    id_to_pending_data_.emplace(sensor_id, timed_point_cloud_data);
+    return result;
+  }
+  id_to_pending_data_.emplace(sensor_id, timed_point_cloud_data);
+  if (expected_sensor_ids_.size() == id_to_pending_data_.size()) {
```
Comment (pifon2a) [inline cartographer/mapping/internal/range_data_collator.cc]: flip it:
```
if (expected_sensor_ids_.size() != id_to_pending_data_.size()) return {};
current_start_...
```

File: cartographer/mapping/internal/range_data_collator.cc (hunk 2 above)
Comment (gaschler) [reply cartographer/mapping/internal/range_data_collator.cc]: Very unlikely to happen. But if it happens, we should really force-stop, so I would like to keep this check.

File: cartographer/mapping/internal/3d/local_trajectory_builder_3d_test.cc (hunk 1 above)
Comment (gaschler) [reply cartographer/mapping/internal/3d/local_trajectory_builder_3d_test.cc]: Done

File: cartographer/mapping/internal/range_data_collator.cc (hunk 3 above)
Comment (gaschler) [reply cartographer/mapping/internal/range_data_collator.cc]: Done

**Comment by gaschler** [comment]
Splitting up this PR in smaller ones, see https://github.com/googlecartographer/cartographer/pull/975, closing.

**Pull request closed without merging**